iml-manager-cli/src/update_repo_file.rs

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) 2019 DDN. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+use crate::{
+    api_utils::{create_command, get_hosts, run_cmd, wait_for_cmd, SendCmd, SendJob},
+    display_utils::{display_cancelled, display_cmd_state, display_error, start_spinner},
+    error::ImlManagerCliError,
+};
+use futures::{
+    future::{self, loop_fn, Either, Loop},
+    Future,
+};
+use iml_wire_types::Host;
+use std::collections::BTreeSet;
+use structopt::StructOpt;
+
+#[derive(StructOpt, Debug)]
+pub struct UpdateRepoFileHosts {
+    /// The host(s) to update. Takes a hostlist expression
+    #[structopt(short = "h", long = "hosts")]
+    hosts: String,
+}
+
+fn filter_known_hosts<'a, 'b>(
+    hostlist: &'b BTreeSet<String>,
+    api_hosts: &'a Vec<Host>,
+) -> Vec<&'a Host> {
+    api_hosts
+        .iter()
+        .filter(move |x| hostlist.contains(&x.fqdn) || hostlist.contains(&x.nodename))
+        .collect()
+}
+
+fn has_host(api_hosts: &Vec<&Host>, host: &String) -> bool {
+    api_hosts
+        .iter()
+        .any(|y| &y.fqdn == host || &y.nodename == host)
+}
+
+#[derive(serde::Serialize)]
+pub struct HostId {
+    host_id: u32,
+}
+
+pub fn update_repo_file_cli(config: UpdateRepoFileHosts) -> Result<(), ImlManagerCliError> {
+    let r = hostlist_parser::parse(&config.hosts).expect("Couldn't parse hostlist expression.");
+
+    tracing::debug!("Parsed hosts {:?}", r);
+
+    let stop_spinner = start_spinner("Fetching hosts...");
+
+    let fut: impl Future<Item = Command, Error = ImlManagerCliError> = get_hosts()
+        .and_then(move |api_hosts| {
+            stop_spinner(None);
+
+            let xs = filter_known_hosts(&r, &api_hosts.objects);
+
+            for x in r.iter() {
+                if !has_host(&xs, &x) {
+                    display_cancelled(format!("Host {} unknown to IML.", x));
+                }
+            }
+
+            if xs.is_empty() {
+                display_error("0 hosts can be updated. Exiting");
+
+                return Ok(());
+            }
+
+            let cmd = SendCmd {
+                jobs: xs
+                    .iter()
+                    .map(|x| SendJob {
+                        class_name: "UpdateYumFileJob".into(),
+                        args: HostId { host_id: x.id },
+                    })
+                    .collect(),
+                message: "Update IML Agent Repo files".into(),
+            };
+
+            let stop_spinner = start_spinner("Updating Repo files...");
+
+            create_command(cmd).map(|cmd| (cmd, stop_spinner))
```

Comment:
I'm having an issue here... I should be able to return a future but it only lets me return `()`